### PR TITLE
When creating a new flex project via the New Project Wizard

### PR DESCRIPTION
**Pull request by nbashirbello**
- install the flex facet, Java 8 facet and web facet in the sample project
- download the dependencies of the sample project from maven and add it to it

### CODE REVIEW

**Comment by nbashirbello** [comment]
Fixes #1577

File: plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+ * Tests for {@code FacetUtil}
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class FacetUtilTest {
+  @Mock private IFacetedProject mockFacetedProject;
+  @Rule public TestProjectCreator projectCreator = new TestProjectCreator();
+  private Set<IFacetedProject.Action> facetInstallSet = new HashSet<>();
+
+  @Test
+  public void TestAddJavaFacetToBatch_facetDoesNotExitsInProject() throws CoreException
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtilTest.java]: method names should start with lowercase.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacet.java [hunk 2]
```diff
@@ -32,4 +42,35 @@ public static boolean hasAppEngineFacet(IFacetedProject project) {
     return FacetedProjectHelper.projectHasFacet(project, ID);
   }
 
+  /**
+   * Checks to see if <code>facetedProject</code> has the App Engine Flexible facet.
+   * If not, it installs the App Engine Flexible facet.
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacet.java]: I'd rephrase it to 
```Installs the App Engine Flexible facet if {@code facetedProject} does not have it already.```

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacet.java [hunk 3]
```diff
@@ -32,4 +42,35 @@ public static boolean hasAppEngineFacet(IFacetedProject project) {
...
+  public static void installAppEngineFacet(IFacetedProject facetedProject,
+      boolean installDependentFacets, IProgressMonitor monitor) throws CoreException {
+    Set<IFacetedProject.Action> facetInstallSet = new HashSet<>();
+
+    IProjectFacet appEngineFacet = ProjectFacetsManager.getProjectFacet(AppEngineFlexFacet.ID);
+    if (facetedProject.hasProjectFacet(appEngineFacet)) {
+      return;
+    }
+
+    // Get the Java 8 facet
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacet.java]: This comment in unnecessary.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacet.java [hunk 4]
```diff
@@ -32,4 +42,35 @@ public static boolean hasAppEngineFacet(IFacetedProject project) {
...
+    if (installDependentFacets) {
+      FacetUtil.addJavaFacetToBatch(JavaFacet.VERSION_1_8, facetedProject, facetInstallSet);
+    }
+
+    // Get App Engine facet version
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineFlexFacet.java]: This comment in unnecessary.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java [hunk 5]
```diff
@@ -142,8 +128,8 @@ public static void installAppEngineFacet(IFacetedProject facetedProject,
     Set<IFacetedProject.Action> facetInstallSet = new HashSet<>();
     // Install required App Engine facets i.e. Java 1.7 and Dynamic Web Module 2.5
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java]: This comment in unnecessary.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java [hunk 6]
```diff
@@ -58,6 +42,7 @@
 import org.eclipse.wst.server.core.IRuntimeWorkingCopy;
 import org.eclipse.wst.server.core.ServerCore;
 
+@SuppressWarnings("restriction") // For DependencyGraphImpl and IDependencyGraph
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineStandardFacet.java]: can this be moved to the method these warnings appear to suppress them?

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+import org.eclipse.wst.common.componentcore.internal.builder.IDependencyGraph;
+import org.eclipse.wst.common.frameworks.datamodel.DataModelFactory;
+import org.eclipse.wst.common.frameworks.datamodel.IDataModel;
+import org.eclipse.wst.common.project.facet.core.IFacetedProject;
+import org.eclipse.wst.common.project.facet.core.IProjectFacetVersion;
+
+/**
+ * Utility class for processing facets.
+ */
+@SuppressWarnings("restriction") // For IDependencyGraph
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: can this be moved to the method these warnings appear to suppress them?

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+  /**
+   * Adds an install action for the {@code javaFacet} to the {@code facetInstallSet} if the
+   * {@code javaFacet} does not already exist in {@code facetedProject}.
+   *
+   * @return true is the facet action was added to {@code facetInstallSet} and false if {@code javaFacet}
+   *   already exists in {@code facetedProject}
+   * @throws CoreException if {@code javaFacet} is not a Java facet
+   */
+ public static boolean addJavaFacetToBatch(IProjectFacetVersion javaFacet, IFacetedProject facetedProject,
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: Instead of passing in an `IProjectFacetVersion` and check if it's a Java facet, we could just pass in the version we want to add (possibly defined in an enum to restrict the possible values) and get rid of the facet ID check 2 lines below.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+     Set<IFacetedProject.Action> facetInstallSet) throws CoreException {
+   if (!JavaFacet.FACET.getId().equals(javaFacet.getProjectFacet().getId())) {
+     throw new CoreException(StatusUtil.error(FacetUtil.class, javaFacet.toString() + " is not a Java facet"));
+   }
+   
+   if (facetedProject.hasProjectFacet(javaFacet)) {
+     return false;
+   }
+
+   // TODO use "src/main/java" for only maven projects
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: This comment should appear about 4 lines below, and a link to the corresponding issue would be useful: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/490

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+
+ /**
+  * Adds an install action for the {@code webFacet} to the {@code facetInstallSet} if the
+  * {@code webFacet} does not already exist in {@code facetedProject}.
+  * 
+  * @return true is the facet action was added to {@code facetInstallSet} and false if {@code webFacet}
+  *   already exists in {@code facetedProject}
+  * @throws CoreException if {@code webFacet} is not a Web facet
+  */
+ public static boolean addWebFacetToBatch(IProjectFacetVersion webFacet, IFacetedProject facetedProject,
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: Similar to adding Java facet, we could use only version for web facet as parameter.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java [hunk 11]
```diff
@@ -16,28 +16,71 @@
...
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jst.common.project.facet.core.JavaFacet;
+import org.eclipse.jst.j2ee.web.project.facet.WebFacetUtils;
+import org.eclipse.wst.common.project.facet.core.IFacetedProject;
+import org.eclipse.wst.common.project.facet.core.ProjectFacetsManager;
 
 /**
-* Utility to make a new Eclipse project with the App Engine Flexible facets in the workspace.
-*/
+ * Utility to make a new App Engine Flexible Eclipse project.
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: `create` instead of `make`?

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java [hunk 12]
```diff
@@ -16,28 +16,71 @@
...
+ */
 public class CreateAppEngineFlexWtpProject extends CreateAppEngineWtpProject {
+  private static final Logger logger = Logger.getLogger(CreateAppEngineFlexWtpProject.class.getName());
+  private static final Map<String, String> PROJECT_DEPENDENCIES;
+  static {
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: blank line above

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java [hunk 13]
```diff
@@ -46,9 +89,88 @@ public String getDescription() {
...
+    SubMonitor subMonitor = SubMonitor.convert(monitor, 100);
+
+    // Create a lib folder
+    IFolder libFolder = project.getFolder("lib");
+    if (!libFolder.exists()) {
+      libFolder.create(true, true, subMonitor.newChild(10));
+    }
+
+    // Download the dependencies from maven
+    M2RepositoryService repoService = new M2RepositoryService();
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: `M2RepositoryService` is available as an OSGi service through the `ILibraryRepositoryService` service interface, please use that instance instead of creating a new one.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java [hunk 14]
```diff
@@ -46,9 +89,88 @@ public String getDescription() {
...
+    repoService.activate();
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: This will be called by the OSGi framework (if needed) when accessing the service, so no need to call it.

File: plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/repository/M2RepositoryService.java [hunk 15]
```diff
@@ -98,7 +98,7 @@ private static IPath getDownloadedSourceLocation(MavenCoordinates mavenCoordinat
   }
 
   @Activate
-  protected void activate() {
+  public void activate() {
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/repository/M2RepositoryService.java]: You should use the OGSi service instance of `M2RepositoryService` to access it. And keep this method protected.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java [hunk 16]
```diff
@@ -46,9 +89,88 @@ public String getDescription() {
...
+    IClasspathEntry[] entries = javaProject.getRawClasspath();
+    List<IClasspathEntry> newEntries = new ArrayList<IClasspathEntry>();
+    newEntries.addAll(Arrays.asList(entries));
+
+    // Add all the jars under lib folder to the classpath
+    File libFolder = new File(libraryPath);
+
+    for(File file : libFolder.listFiles()) {
+      IPath path = Path.fromOSString(file.toPath().toString());
+      newEntries.add(JavaCore.newLibraryEntry(path, null, null));
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: We should attach the sources and javadocs if available. Probably in a separate PR though.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtilTest.java [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+    createPath(project, new Path("webapps/first-webapp/WEB-INF"));
+    createEmptyFile(project, new Path("webapps/second-webapp/WEB-INF/web.xml"));
+    createPath(project, new Path("third-webapp/WEB-INF"));
+    createPath(project, new Path("WEB-INF"));
+
+    IPath mainWebApp = FacetUtil.findMainWebAppDirectory(project);
+    Assert.assertEquals(new Path("webapps/second-webapp"), mainWebApp);
+  }
+
+  private static void createEmptyFile(IProject project, IPath relativePath) throws CoreException {
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtilTest.java]: Please add a test for the multiple web.xml case.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtilTest.java [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+    createPath(project, relativePath.removeLastSegments(1));
+
+    InputStream emptyStream = new ByteArrayInputStream(new byte[0]);
+    project.getFile(relativePath).create(emptyStream, false /* force */, null /* monitor */);
+    Assert.assertTrue(project.getFile(relativePath).exists());
+  }
+
+  private static void createPath(IContainer parent, IPath relativePath) throws CoreException {
```
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets.test/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtilTest.java]: Add a reference to https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1573 and add this method to the issue's description so we can keep track of it.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java (hunk 8 above)
Comment (nbashirbello) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: I rather not create a enum for another projects variable when there is a way to verify the validity of the javaFacet.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java (hunk 10 above)
Comment (nbashirbello) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: Same as above.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+   * Adds an install action for the {@code javaFacet} to the {@code facetInstallSet} if the
+   * {@code javaFacet} does not already exist in {@code facetedProject}.
+   *
+   * @return true is the facet action was added to {@code facetInstallSet} and false if {@code javaFacet}
+   *   already exists in {@code facetedProject}
+   * @throws CoreException if {@code javaFacet} is not a Java facet
+   */
+ public static boolean addJavaFacetToBatch(IProjectFacetVersion javaFacet, IFacetedProject facetedProject,
+     Set<IFacetedProject.Action> facetInstallSet) throws CoreException {
+   if (!JavaFacet.FACET.getId().equals(javaFacet.getProjectFacet().getId())) {
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: Would it make sense for this to be a `Preconditions` assertion?

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+import org.eclipse.wst.common.frameworks.datamodel.DataModelFactory;
+import org.eclipse.wst.common.frameworks.datamodel.IDataModel;
+import org.eclipse.wst.common.project.facet.core.IFacetedProject;
+import org.eclipse.wst.common.project.facet.core.IProjectFacetVersion;
+
+/**
+ * Utility class for processing facets.
+ */
+@SuppressWarnings("restriction") // For IDependencyGraph
+public class FacetUtil {
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: I think it would be worth remaking this class into more of a fluent-style API:

```
FacetUtil.configure(facetedProject)
.addJavaFacet(…)
.addWebFacet(…)
.addFacet(…)
.install();
```

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+     Set<IFacetedProject.Action> facetInstallSet) throws CoreException {
+   if (!WebFacetUtils.WEB_FACET.getId().equals(webFacet.getProjectFacet().getId())) {
+     throw new CoreException(StatusUtil.error(FacetUtil.class, webFacet.toString() + " is not a Web facet"));
+   }
+
+   if (facetedProject.hasProjectFacet(webFacet)) {
+     return false;
+   }
+
+   String webAppDirectory = "src/main/webapp";
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: It would be good to cause`WebProjectUtil.getWebInfFolder()` to share this functionality.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java [hunk 22]
```diff
@@ -16,28 +16,76 @@
...
 
   @Override
   public void addAppEngineFacet(IProject newProject, IProgressMonitor monitor) throws CoreException {
-    // TODO add flex facet
+    SubMonitor subMonitor = SubMonitor.convert(monitor,
+        Messages.getString("add.appengine.flex.facet"), 100);
+
+    IFacetedProject facetedProject = ProjectFacetsManager.create(
+        newProject, true /* convertIfNecessary */, subMonitor.newChild(50));
+    AppEngineFlexFacet.installAppEngineFacet(
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: It seems odd to have this call alongside `configureFacets()`. Each call to `IFacetedProject.modify()` triggers the Dependency Graph job, and more.  Perhaps this could take the install `Set<>`?

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java [hunk 23]
```diff
@@ -46,9 +94,86 @@ public String getDescription() {
...
+
+      // Copy dependency from local maven repo into lib folder
+      if (artifact != null) {
+        File artifactFile = artifact.getFile();
+        IFile destFile = libFolder.getFile(artifactFile.getName());
+        try {
+          destFile.create(new FileInputStream(artifactFile), true, subMonitor.newChild(30));
+        } catch (FileNotFoundException ex) {
+          logger.log(Level.WARNING, "Error copying over " + artifactFile.toString() + " to " +
+        libFolder.getFullPath().toPortableString(), ex);
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: Nit: indentation is off.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java [hunk 24]
```diff
@@ -46,9 +94,86 @@ public String getDescription() {
...
+      libraryFile.getMavenCoordinates().toString() + " from maven", ex);
+        continue;
+      }
+
+      // Copy dependency from local maven repo into lib folder
+      if (artifact != null) {
+        File artifactFile = artifact.getFile();
+        IFile destFile = libFolder.getFile(artifactFile.getName());
+        try {
+          destFile.create(new FileInputStream(artifactFile), true, subMonitor.newChild(30));
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: Ah, I just learned that `create()` always closes input stream.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java [hunk 25]
```diff
@@ -16,28 +16,76 @@
...
-*/
+ * Utility to create a new App Engine Flexible Eclipse project.
+ */
 public class CreateAppEngineFlexWtpProject extends CreateAppEngineWtpProject {
+  private static final Logger logger = Logger.getLogger(CreateAppEngineFlexWtpProject.class.getName());
+  private static final Map<String, String> PROJECT_DEPENDENCIES;
+
+  static {
+    Map<String, String> projectDependencies = new HashMap<String, String>();
+    projectDependencies.put("javax.servlet", "servlet-api");
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: Interestingly, it adds `servlet-api-3.0-alpha-1.jar` to the project. I think its version should be pinned to be `3.0.1` to get away from that `alpha` tag.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java [hunk 26]
```diff
@@ -46,9 +94,86 @@ public String getDescription() {
...
+      // Copy dependency from local maven repo into lib folder
+      if (artifact != null) {
+        File artifactFile = artifact.getFile();
+        IFile destFile = libFolder.getFile(artifactFile.getName());
+        try {
+          destFile.create(new FileInputStream(artifactFile), true, subMonitor.newChild(30));
+        } catch (FileNotFoundException ex) {
+          logger.log(Level.WARNING, "Error copying over " + artifactFile.toString() + " to " +
+        libFolder.getFullPath().toPortableString(), ex);
+          continue;
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: Nit: this is not necessary.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java [hunk 27]
```diff
@@ -46,9 +94,86 @@ public String getDescription() {
...
+
+    // Download the dependencies from maven
+    int ticks = 50 / PROJECT_DEPENDENCIES.size();
+    for (Map.Entry<String, String> dependency : PROJECT_DEPENDENCIES.entrySet()) {
+      LibraryFile libraryFile = new LibraryFile(new MavenCoordinates(dependency.getKey(),
+          dependency.getValue()));
+      Artifact artifact = null;
+      try {
+        artifact = repositoryService.resolveArtifact(libraryFile, subMonitor.newChild(ticks));
+      } catch (CoreException ex) {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: Looking at this logic, I believe it's better to merge this `try` block and the other block below IMO: it's either that you succeed to copy the artifact or not. (Basically, I am talking about Advantage 1 of the exception mechanism in [here](http://docs.oracle.com/javase/tutorial/essential/exceptions/advantages.html).) What you are doing now seems just like the traditional error code checking (i.e., using `if` to check a return code of a function).

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java (hunk 25 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: Hmm, good find: `javax.servlet:servlet-api` looks to be [ancient](http://search.maven.org/#search%7Cga%7C1%7Cg%3A%22javax.servlet%22), last updated in 2008.  Looks like it should be `javax.servlet:javax.servlet-api` and we should definitely pin a version since 4.0 is under development.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java (hunk 20 above)
Comment (nbashirbello) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java]: Done

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java (hunk 25 above)
Comment (nbashirbello) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1674

Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: I think we should do this in this PR, though not 4.0. 3.1 is right for flex:

https://cloud.google.com/appengine/docs/flexible/

Comment (nbashirbello) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/flex/CreateAppEngineFlexWtpProject.java]: The updating of the jar requires us to update the MockHttpServletResponse, which means I have to change the code templates and all. Since after the large team meeting we are trying to keep the PRs smaller and more focused, I thought it was best to update the templates in another PR. Your call though.
